couchbase_test.go

```diff
@@ -0,0 +1,775 @@
+package couchbase
+
+import (
+	"context"
+	"fmt"
+	"github.com/cenkalti/backoff"
+	docker "github.com/hashicorp/vault/helper/testhelpers/docker"
+	"github.com/hashicorp/vault/sdk/database/dbplugin"
+	"github.com/ory/dockertest"
+	dc "github.com/ory/dockertest/docker"
+	"net/http"
+	"os"
+	"testing"
+	"time"
+)
+
+var pre6dot5 = false // check for Pre 6.5.0 Couchbase
+
+const (
+	adminUsername = "Administrator"
+	adminPassword = "password"
+	bucketName    = "travel-sample"
+)
+
+func prepareCouchbaseTestContainer(t *testing.T) (func(), string, int) {
+	if os.Getenv("COUCHBASE_HOST") != "" {
+		return func() {}, os.Getenv("COUCHBASE_HOST"), 8091
+	}
+	// cbver should match a couchbase/server-sandbox docker repository tag. Default to 6.5.0
+	cbver := os.Getenv("COUCHBASE_VERSION")
+	if cbver == "" {
+		cbver = "6.5.0"
+	}
+
+	pool, err := dockertest.NewPool("")
+	if err != nil {
+		t.Fatalf("Failed to connect to docker: %s", err)
+	}
+
+	ro := &dockertest.RunOptions{
+		Repository:   "docker.io/couchbase/server-sandbox",
+		Tag:          cbver,
+		ExposedPorts: []string{"8091", "8092", "8093", "8094", "11207", "11210", "18091", "18092", "18093", "18094"},
+		PortBindings: map[dc.Port][]dc.PortBinding{
+			"8091": {
+				{HostIP: "0.0.0.0", HostPort: "8091"},
+			},
+			"8092": {
+				{HostIP: "0.0.0.0", HostPort: "8092"},
+			},
+			"8093": {
+				{HostIP: "0.0.0.0", HostPort: "8093"},
+			},
+			"8094": {
+				{HostIP: "0.0.0.0", HostPort: "8094"},
+			},
+			"11207": {
+				{HostIP: "0.0.0.0", HostPort: "11207"},
+			},
+			"11210": {
+				{HostIP: "0.0.0.0", HostPort: "11210"},
+			},
+			"18091": {
+				{HostIP: "0.0.0.0", HostPort: "18091"},
+			},
+			"18092": {
+				{HostIP: "0.0.0.0", HostPort: "18092"},
+			},
+			"18093": {
+				{HostIP: "0.0.0.0", HostPort: "18093"},
+			},
+			"18094": {
+				{HostIP: "0.0.0.0", HostPort: "18094"},
+			},
+		},
+	}
+	resource, err := pool.RunWithOptions(ro)
+	if err != nil {
+		t.Fatalf("Could not start local couchbase docker container: %s", err)
+	}
+
+	cleanup := func() {
+		docker.CleanupResource(t, pool, resource)
+	}
+
+	address := "http://127.0.0.1:8091/"
+	
+	if err = pool.Retry(func() error {
+		t.Log("Waiting for the database to start...")
+		resp, err := http.Get(address)
+		if err != nil {
+			return err
+		}
+		if resp.StatusCode != 200 {
+			return fmt.Errorf("Got a %d status code from couchbase's Web Console", resp.StatusCode)
+		}
+		return nil
+	}); err != nil {
+		t.Fatalf("Could not connect to couchbase: %s", err)
```

Comment:
I added the cleanup call. I can't see what Circleci is doing and I can't run it myself any more as I get this error due to adding docker layer caching on a free plan.

This job has been blocked because Docker Layer Caching is not available on your current plan. Please upgrade to continue building.

Reply:
Hi Francis, I pushed your branch (fhitchen-master) to our main repo and you should be able to see the build run here - https://app.circleci.com/pipelines/github/hashicorp/vault-plugin-database-couchbase/6/workflows/54f05ed7-aad3-4d5b-9961-0c1199080563/jobs/13 - it looks like it failed with the same error as @tomhjp mentioned. This appears to be intermittent and likely will pass on the next few runs, so just FYI.

Reply:
Hi Sarah, do you know what resources the VM's that are running the CicrcleCI jobs have assigned to them? I used Github workflows on another project that required a Couchbase DB up and running to execute the integration tests and I had to use these docker options to increase CPU and memory for the docker container in order to get consistent stable tests to run ...
```bash
    services:
      couchbase:
        image: docker.io/fhitchen/clj-cb-testdb
        ports:
          - 8091:8091
          - 8092:8092
          - 8093:8093
          - 8094:8094
          - 8095:8095
          - 8096:8096
          - 11210:11210
          - 11211:11211
        options: --cpus 2 --memory 4GB
```
Couchbase gets flaky if you use just one core and not much memory.

Reply:
Hi Francis, I added the `resource_class: large` setting to the CircleCI yaml and had got a [couple of green runs](https://app.circleci.com/pipelines/github/hashicorp/vault-plugin-database-couchbase). This is [4 vCPUS and 15GB RAM](https://circleci.com/docs/2.0/configuration-reference/#machine-executor-linux). Although by default it would have been using the medium resource class which would have been 2 vCPUS and 7.5GB of RAM.

Reply:
Thanks for testing that @sarahethompson.

Given the existing tests have satisfied the [minimum specifications](https://docs.couchbase.com/server/current/install/pre-install.html) from the docs (and by default, docker runs without resources limits, so it will have had access to the whole VM), I think resources constraints are an unlikely candidate.

However, given the low failure rate, I don't see much value in chasing this down, unless you're particularly keen to @ fhitchen. I'm happy to merge this PR if you're ready for me to?

Reply:
Yes, please go ahead and merge. I have seen it fail on my old under powered Linux server, but it is not repeatable. The only other thing I can think it could be if it is not the server resources, is the timeout on the UpsertUser calls. It is set at 5 seconds, I will try and reproduce and see if upping this default is the cure.